pkg/controller/mattermostcluster/mattermostcluster_controller.go

```diff
@@ -100,54 +110,524 @@ func (r *ReconcileMattermostCluster) Reconcile(request reconcile.Request) (recon
 		return reconcile.Result{}, err
 	}
 
-	// Define a new Pod object
-	pod := newPodForCR(instance)
-
-	// Set MattermostCluster instance as the owner and controller
-	if errSetController := controllerutil.SetControllerReference(instance, pod, r.scheme); errSetController != nil {
-		return reconcile.Result{}, errSetController
+	err = r.setDefaults(mattermost, reqLogger)
+	if err != nil {
+		return reconcile.Result{}, err
 	}
 
-	// Check if this Pod already exists
-	found := &corev1.Pod{}
-	err = r.client.Get(context.TODO(), types.NamespacedName{Name: pod.Name, Namespace: pod.Namespace}, found)
-	if err != nil && errors.IsNotFound(err) {
-		reqLogger.Info("Creating a new Pod", "Pod.Namespace", pod.Namespace, "Pod.Name", pod.Name)
-		err = r.client.Create(context.TODO(), pod)
+	if mattermost.Spec.DatabaseType.Type == "mysql" {
+		reqLogger.Info("Reconciling MattermostCluster Database MySQL service account")
+		err = r.checkDBServiceAccount(mattermost, reqLogger)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+
+		reqLogger.Info("Reconciling MattermostCluster Database MySQL role binding")
+		err = r.checkDBRoleBinding(mattermost, reqLogger)
 		if err != nil {
 			return reconcile.Result{}, err
 		}
 
-		// Pod created successfully - don't requeue
-		return reconcile.Result{}, nil
-	} else if err != nil {
+		reqLogger.Info("Reconciling MattermostCluster Database MySQL")
+		err = r.checkDBMySQLDeployment(mattermost, reqLogger)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+	} else {
+		reqLogger.Info("Reconciling MattermostCluster Database Postgres")
+		err = r.checkDBPostgresDeployment(mattermost, reqLogger)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+	}
+
+	// TODO
+	err = r.checkMinioDeployment(mattermost, reqLogger)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	err = r.checkMattermostDeployment(mattermost, reqLogger)
+	if err != nil {
 		return reconcile.Result{}, err
 	}
 
-	// Pod already exists - don't requeue
-	reqLogger.Info("Skip reconcile: Pod already exists", "Pod.Namespace", found.Namespace, "Pod.Name", found.Name)
+	// Already exists - don't requeue
 	return reconcile.Result{}, nil
 }
 
-// newPodForCR returns a busybox pod with the same name/namespace as the cr
-func newPodForCR(cr *mattermostv1alpha1.MattermostCluster) *corev1.Pod {
-	labels := map[string]string{
-		"app": cr.Name,
+func (r *ReconcileMattermostCluster) setDefaults(mattermost *mattermostv1alpha1.MattermostCluster, reqLogger logr.Logger) error {
```

Comment:
This function doesn't actually use the pointer receiver `r *ReconcileMattermostCluster` anywhere so we can just make it a function in the package. I'd also suggest generalizing the function so that it doesn't have to be tied to a reconcile request. That way if someone wants to use this package for use outside the operator they can do so. In fact, it might even make sense to move this into the `v1alpha1` package as a method directly on the `*MattermostCluster` type. So you would just do something like this above in `Reconcile`:

```go
mattermost := &mattermostv1alpha1.MattermostCluster{}
// ... other stuff ...
reqLogger.Info("Setting defaults")
err := mattermost.SetDefaults()
if err != nil {
	return reconcile.Result{}, err
}
```

If we really wanted to keep the `reqLogger` around to log when a default was set, we could then have a function in this package that compares the changes made to the `MattermostCluster` then outputs what defaults were set but at this point I'm not sure if that's needed.

Reply:
done